File: app/src/main/java/com/dss886/dotaautochess/widget/AppHeaderBehavior.java
package com.dss886.dotaautochess.widget;

import android.content.Context;
import android.util.AttributeSet;
import android.view.View;

import com.dss886.dotaautochess.R;
import com.google.android.material.appbar.AppBarLayout;

import androidx.annotation.NonNull;
import androidx.coordinatorlayout.widget.CoordinatorLayout;

/**
 * Created by dss886 on 2019/1/26.
 */
@SuppressWarnings("unused")
public class AppHeaderBehavior extends CoordinatorLayout.Behavior<AppHeaderView> {

    private float mHeaderMoveDistance;

    public AppHeaderBehavior() {}

    public AppHeaderBehavior(Context context, AttributeSet attrs) {
        super(context, attrs);
        mHeaderMoveDistance = context.getResources().getDimension(R.dimen.hero_header_height)
                - context.getResources().getDimension(R.dimen.toolbar_height);
    }

    @Override
    public boolean layoutDependsOn(@NonNull CoordinatorLayout parent, @NonNull AppHeaderView child, @NonNull View dependency) {
        return dependency instanceof AppBarLayout;
    }

    @Override
    public boolean onDependentViewChanged(@NonNull CoordinatorLayout parent, @NonNull AppHeaderView child, @NonNull View dependency) {
        if (mHeaderMoveDistance == 0) {
            return false;
        }
        float percent = -dependency.getY() / mHeaderMoveDistance;
        if (percent < 0) {
            percent = 0;
        }
        child.onCollapsing(percent);
        return true;
    }
}

File: app/src/main/java/com/dss886/dotaautochess/widget/AppHeaderBehavior.kt
package com.dss886.dotaautochess.widget

import android.content.Context
import android.util.AttributeSet
import android.view.View
import androidx.coordinatorlayout.widget.CoordinatorLayout
import com.dss886.dotaautochess.R
import com.google.android.material.appbar.AppBarLayout

/**
 * Created by dss886 on 2019/1/26.
 */
@Suppress("unused", "UNUSED_PARAMETER")
class AppHeaderBehavior @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null) : CoordinatorLayout.Behavior<AppHeaderView>() {

    private val mHeaderMoveDistance = context.resources.getDimension(R.dimen.hero_header_height) - context.resources.getDimension(R.dimen.toolbar_height)

    override fun layoutDependsOn(parent: CoordinatorLayout, child: AppHeaderView, dependency: View): Boolean {
        return dependency is AppBarLayout
    }

    override fun onDependentViewChanged(parent: CoordinatorLayout, child: AppHeaderView, dependency: View): Boolean {
        if (mHeaderMoveDistance == 0f) {
            return false
        }
        var percent = -dependency.y / mHeaderMoveDistance
        if (percent < 0) {
            percent = 0f
        }
        child.onCollapsing(percent)
        return true
    }
}
